### PR TITLE
FileChooser portal

### PR DESCRIPTION
**Pull request by jackpot51**
Requires https://github.com/pop-os/cosmic-files/pull/269

Known issues (that I do not consider blocking):
- The SaveFiles dialog does not check for overwriting files and instead just uses the open folder dialog
- Attaching the dialog to a window is not implemented
- The modal setting is not used at all

### CODE REVIEW

**Review by mmstick — APPROVED** [review]

File: src/file_chooser.rs [hunk 1]
```diff
@@ -0,0 +1,474 @@
...
+                                    if *value { "true" } else { "false" }.to_string(),
+                                ));
+                            }
+                            DialogChoice::ComboBox {
+                                id,
+                                options,
+                                selected,
+                                ..
+                            } => {
+                                //TODO: what to fill in if selected option not set?
```
Comment (wash2) [inline src/file_chooser.rs]: https://github.com/flatpak/xdg-desktop-portal/blob/main/data/org.freedesktop.portal.FileChooser.xml implies that the portal can choose the initially selected value if no value is provided in the request, so maybe just picking the first option would be fine as a default.

File: src/file_chooser.rs [hunk 2]
```diff
@@ -0,0 +1,474 @@
...
+            let response = match dialog_res {
+                DialogResult::Cancel => PortalResponse::Cancelled,
+                DialogResult::Open(paths) => {
+                    let mut uris = Vec::with_capacity(paths.len());
+                    for path in paths {
+                        match url::Url::from_file_path(&path) {
+                            Ok(url) => uris.push(url.to_string()),
+                            Err(()) => {
+                                log::error!("failed to convert to URL: {:?}", path);
+                                //TODO: return error?
```
Comment (wash2) [inline src/file_chooser.rs]: If `uris` is empty after this step, maybe `PortalResponse::Other` could be returned?

File: src/file_chooser.rs [hunk 3]
```diff
@@ -0,0 +1,474 @@
...
+        tokio::spawn(async move {
+            let _ = old_args
+                .tx
+                .send(PortalResponse::Cancelled::<FileChooserResult>)
+                .await;
+        });
+    }
+
+    if let Some(old_dialog) = portal.file_chooser_dialog.take() {
+        //TODO: close window
```
[the remark this answers is not in the document]
Comment (jackpot51) [reply src/file_chooser.rs]: This code is no longer used after switching to multiple file dialogs

File: src/file_chooser.rs (hunk 2 above)
Comment (jackpot51) [reply src/file_chooser.rs]: Fixed

File: src/file_chooser.rs (hunk 1 above)
Comment (jackpot51) [reply src/file_chooser.rs]: Fixed

**Review by wash2 — APPROVED** [review]

**Pull request merged**